<!DOCTYPE html>

<html lang="en">
<head>
<meta charset="utf-8"/>
<meta content="IE=edge" http-equiv="X-UA-Compatible"/>
<meta content="width=device-width, initial-scale=1.0" name="viewport"/>
<meta content="None" name="description"/>
<link href="img/favicon.ico" rel="shortcut icon"/>
<title>SCALLOP-INF</title>
<link href="css/bootstrap.min.css" rel="stylesheet"/>
<link href="css/font-awesome.min.css" rel="stylesheet"/>
<link href="css/base.css" rel="stylesheet"/>
<link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css" rel="stylesheet"/>
<script defer="" src="js/jquery-1.10.2.min.js"></script>
<script defer="" src="js/bootstrap.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/Bash.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/HTML.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/Java.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/JavaScript.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/Perl.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/Python.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/R.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/rust.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/yaml.min.js"></script>
<script>hljs.initHighlightingOnLoad();</script>
</head>
<body class="homepage">
<div class="navbar fixed-top navbar-expand-lg navbar-dark bg-dark">
<div class="container">
<a class="navbar-brand" href=".">SCALLOP-INF</a>
<!-- Expander button -->
<button class="navbar-toggler" data-target="#navbar-collapse" data-toggle="collapse" type="button">
<span class="navbar-toggler-icon"></span>
</button>
<!-- Expanded navigation -->
<div class="navbar-collapse collapse" id="navbar-collapse">
<!-- Main navigation -->
<ul class="nav navbar-nav">
<li class="navitem active">
<a class="nav-link" href=".">Home</a>
</li>
<li class="navitem">
<a class="nav-link" href="doc/SCALLOP_INF1_analysis_plan/">Analysis plan</a>
</li>
<li class="dropdown">
<a class="nav-link dropdown-toggle" data-toggle="dropdown" href="#">Implementation <b class="caret"></b></a>
<ul class="dropdown-menu">
<li>
<a class="dropdown-item" href="tryggve/">Tryggve</a>
</li>
<li>
<a class="dropdown-item" href="cardio/">Cardio</a>
</li>
<li>
<a class="dropdown-item" href="csd3/">CSD3</a>
</li>
</ul>
</li>
<li class="navitem">
<a class="nav-link" href="rsid/">Notes</a>
</li>
</ul>
<ul class="nav navbar-nav ml-auto">
<li class="nav-item">
<a class="nav-link" data-target="#mkdocs_search_modal" data-toggle="modal" href="#">
<i class="fa fa-search"></i> Search
                            </a>
</li>
<li class="nav-item">
<a class="nav-link disabled" rel="prev">
<i class="fa fa-arrow-left"></i> Previous
                                </a>
</li>
<li class="nav-item">
<a class="nav-link" href="doc/SCALLOP_INF1_analysis_plan/" rel="next">
                                    Next <i class="fa fa-arrow-right"></i>
</a>
</li>
</ul>
</div>
</div>
</div>
<div class="container">
<div class="row">
<div class="col-md-3"><div class="navbar-light navbar-expand-md bs-sidebar hidden-print affix" role="complementary">
<div class="navbar-header">
<button class="navbar-toggler collapsed" data-target="#toc-collapse" data-toggle="collapse" title="Table of Contents" type="button">
<span class="fa fa-angle-down"></span>
</button>
</div>
<div class="navbar-collapse collapse card bg-secondary" id="toc-collapse">
<ul class="nav flex-column">
<li class="nav-item" data-level="1"><a class="nav-link" href="#scallop-inf-meta-analysis">SCALLOP-INF meta-analysis</a>
<ul class="nav flex-column">
<li class="nav-item" data-level="2"><a class="nav-link" href="#flow-of-analysis">Flow of analysis</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#comments">Comments</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#the-opg-example">The OPG example</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#ewas-with-il-12b">EWAS with IL-12B</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#summary-statistics">Summary statistics</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#related-links">Related links</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#references">References</a>
<ul class="nav flex-column">
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div></div>
<div class="col-md-9" role="main">
<h1 id="scallop-inf-meta-analysis">SCALLOP-INF meta-analysis</h1>
<p align="center"><img src="doc/circos.svg"/></p>
<h2 id="flow-of-analysis">Flow of analysis</h2>
<p>(When the diagram is not rendered, view it from <a href="https://mermaid-js.github.io/mermaid-live-editor/">Mermaid live editor</a>)</p>
<div class="mermaid">graph TB;
  tryggve ==&gt; cardio;
  cardio ==&gt; csd3;
  csd3 --&gt; csd3Analysis[Conditional analysis,finemapping, etc];
  csd3 --&gt; software[R Packages at CRAN/GitHub]; 
  tryggveAnalysis[Meta analysis: list.sh, format.sh,metal.sh, QCGWAS.sh, analysis.sh] --&gt; GWAS[pQTL selection and Characterization];
  GWAS --&gt; Prototyping[Prototyping: INTERVAL.sh, cardio.sh, ...];
  Prototyping --&gt; Multi-omics-analysis;
</div>
<h2 id="comments">Comments</h2>
<p>The <a href="tryggve">tryggve</a>, <a href="cardio">cardio</a> and <a href="csd3">csd3</a> directories here are associated with the named Linux cluster(s) used for the analysis over time.</p>
<ol>
<li>
<p>Data pre-processing was done initially from tryggve with <a href="tryggve/list.sh">list.sh</a> and <a href="tryggve/format.sh">format.sh</a>, followed by meta-analysis according to <a href="tryggve/metal.sh">metal.sh</a> using METAL whose results were cross-examined with <a href="tryggve/QCGWAS.sh">QCGWAS.sh</a> together with additional investigation.</p>
</li>
<li>
<p>The main analysis followed with <a href="tryggve/analysis.sh">analysis.sh</a> containing codes for Manhattan/Q-Q/forest/LocusZoom plots such as the OPG example (see the diagram below), which replicated results of Kwan et al. (2014) as identified by PhenoScanner), clumping using PLINK and conditional analysis using GCTA. The clumping results were classified into cis/trans signals. As the meta-analysis stabilised especially with INTERVAL reference, analysis has been intensively done locally with cardio and csd3. cis/trans classification has been done via <a href="cardio/cis.vs.trans.classification.R">cis.vs.trans.classification.R</a> as validated by <a href="cardio/cistrans.sh">cistrans.sh</a>.</p>
</li>
<li>
<p>We prototyped our analysis on cardio with INTERVAL such as <a href="tryggve/INTERVAL.sh">INTERVAL.sh</a> and <a href="cardio/cardio.sh">cardio.sh</a> as well as individual level data analysis for the KORA study. Most analyses were done locally on CSD3.</p>
</li>
<li>
<p>The <code>cis.vs.trans.classification</code>, <code>circos.cis.vs.trans.plot</code> as with <code>cs</code>, <code>log10p</code>, <code>logp</code>, <code>gc.lambda</code>, <code>invnormal</code>, <code>METAL_forestplot</code>, <code>mhtplot.trunc</code>, <code>mhtplot2d</code>, <code>pvalue</code> functions are now part of R/gap at <a href="https://CRAN.R-project.org/package=gap">CRAN</a> and updates such as <code>pqtl2dplot/pqtl2dplotly/pqtl3dplotly</code> are made at <a href="https://github.com/jinghuazhao/R/">GitHub</a>. Some aspects of the downstream analyses links colocalisation and Mendelian randomisation are also available from <a href="https://jinghuazhao.github.io/R/vignettes/gap.html">gap vignette</a> and <a href="https://jinghuazhao.github.io/pQTLtools/articles/index.html">pQTLtools articles</a>. A nested predictive model based on genotype data G, which link with proteins P1, P2, ..., Pn as predictors for outcome y. Alternative routes are T(P)WAS with <a href="http://gusevlab.org/projects/fusion/">fusion_twas</a> and EWAS with <a href="https://jinghuazhao.github.io/EWAS-fusion/">EWAS-fusion</a>.</p>
</li>
</ol>
<div class="mermaid">graph TD;
  G --&gt; P1
  G --&gt; P2
  G --&gt; ...
  G --&gt; Pn
  P1 --&gt; y
  P2 --&gt; y
  ... --&gt; y
  Pn --&gt; y

  SNP["LD reference panel (bed,bim,fam)"] --&gt; |"EWAS reference panel(top1, blup, lasso, enet, bslmm)"| Methylation;
  Methylation --&gt; Protein;
  SNP --&gt; |"GWAS summary statistics (SNP, A1, A2, Z)"| Protein;
</div>
<h2 id="the-opg-example">The OPG example</h2>
<p>This proves to be a positive control. The stacked image below shows Manhattan, Q-Q, forest and LocusZoom plots.</p>
<p align="center"><img src="doc/OPG.png"/></p>
<h2 id="ewas-with-il-12b">EWAS with IL-12B</h2>
<p>(EWAS, joint/conditional) Q-Q and Manhattan plots from <code>ewas-plot.R</code>.</p>
<p align="center"><img src="doc/ewas-plot.png"/></p>
<h2 id="summary-statistics">Summary statistics</h2>
<p>The link will be added here when available.</p>
<h2 id="related-links">Related links</h2>
<ul>
<li><a href="https://github.com/Olink-Proteomics/OlinkRPackage">OlinkAnalyze</a></li>
<li><a href="https://tinyurl.com/shj46ukj">Olink Insights Stat Analysis</a></li>
<li><a href="http://www.scallop-consortium.com/">SCALLOP consortium</a>.</li>
<li><a href="https://www.olink.com/scallop/">Olink location</a>, <a href="https://www.olink.com/question/what-is-npx/">What is NPX</a>, <a href="https://www.olink.com/scallop-f2f-2019/">F2F London meeting</a>, <a href="https://www.olink.com/data-science-workshop-2019/">Data Science Workshop 2019</a>.</li>
<li><a href="https://github.com/lassefolkersen/scallop">GitHub repository</a> for the 2017 <em>PLoS Genetics</em> paper above.</li>
<li><a href="https://secureremote.dtu.dk/vpn/index.html">securecloud</a>.</li>
<li><a href="https://www.olink.com/data-you-can-trust/publications/">Olink publications</a>.</li>
<li><a href="http://www.phpc.cam.ac.uk/ceu/proteins">SomaLogic plasma protein GWAS summary statistics</a>.</li>
<li><a href="https://twc-stanford.shinyapps.io/aging_plasma_proteome/">Aging Plasma Proteome</a> (<a href="https://github.com/lehallib/DEswan">DEswan</a>).</li>
<li><a href="https://genetics.opentargets.org/immunobase">ImmunoBase</a>.</li>
<li><a href="http://www.wwpdb.org/">Worldwide PDB</a></li>
</ul>
<h2 id="references">References</h2>
<p>Choi SW, Mak TS, O'Reilly PF Tutorial: a guide to performing polygenic risk score analyses. <em>Nat Protoc</em> 15, 2759-2772 (2020), <a href="https://github.com/choishingwan/PRSice">GitHub</a> <a href="https://choishingwan.github.io/PRS-Tutorial/">documentation</a>.</p>
<p>Folkersen L, et al. (2017). Mapping of 79 loci for 83 plasma protein biomarkers in cardiovascular disease. <em>PLoS Genetics</em> 13(4), doi.org/10.1371/journal.pgen.1006706.</p>
<p>Kwan JSH, et al. (2014). Meta-analysis of genome-wide association studies identiﬁes two loci associated with circulating osteoprotegerin levels. <em>Hum Mol Genet</em> 23(24): 6684–6693.</p>
<p>Niewczas MA, et al. (2019). A signature of circulating inflammatory proteins and development of end-stage renal disease in diabetes. <em>Nat Med</em>. https://doi.org/10.1038/s41591-019-0415-5</p>
<p>Sun BB, et al. (2018). Genomic atlas of the human plasma proteome. <em>Nature</em> 558: 73–79.</p></div>
</div>
</div>
<footer class="col-md-12">
<hr/>
<p>Documentation built with <a href="https://www.mkdocs.org/">MkDocs</a>.</p>
</footer>
<script>
            var base_url = ".",
                shortcuts = {"help": 191, "next": 78, "previous": 80, "search": 83};
        </script>
<script defer="" src="js/base.js"></script>
<script defer="" src="javascripts/config.js"></script>
<script defer="" src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script defer="" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
<script defer="" src="search/main.js"></script>
<div aria-hidden="true" aria-labelledby="searchModalLabel" class="modal" id="mkdocs_search_modal" role="dialog" tabindex="-1">
<div class="modal-dialog modal-lg">
<div class="modal-content">
<div class="modal-header">
<h4 class="modal-title" id="searchModalLabel">Search</h4>
<button class="close" data-dismiss="modal" type="button"><span aria-hidden="true">×</span><span class="sr-only">Close</span></button>
</div>
<div class="modal-body">
<p>
                    From here you can search these documents. Enter
                    your search terms below.
                </p>
<form>
<div class="form-group">
<input class="form-control" id="mkdocs-search-query" placeholder="Search..." title="Type search term here" type="search"/>
</div>
</form>
<div id="mkdocs-search-results"></div>
</div>
<div class="modal-footer">
</div>
</div>
</div>
</div><div aria-hidden="true" aria-labelledby="keyboardModalLabel" class="modal" id="mkdocs_keyboard_modal" role="dialog" tabindex="-1">
<div class="modal-dialog">
<div class="modal-content">
<div class="modal-header">
<h4 class="modal-title" id="keyboardModalLabel">Keyboard Shortcuts</h4>
<button class="close" data-dismiss="modal" type="button"><span aria-hidden="true">×</span><span class="sr-only">Close</span></button>
</div>
<div class="modal-body">
<table class="table">
<thead>
<tr>
<th style="width: 20%;">Keys</th>
<th>Action</th>
</tr>
</thead>
<tbody>
<tr>
<td class="help shortcut"><kbd>?</kbd></td>
<td>Open this help</td>
</tr>
<tr>
<td class="next shortcut"><kbd>n</kbd></td>
<td>Next page</td>
</tr>
<tr>
<td class="prev shortcut"><kbd>p</kbd></td>
<td>Previous page</td>
</tr>
<tr>
<td class="search shortcut"><kbd>s</kbd></td>
<td>Search</td>
</tr>
</tbody>
</table>
</div>
<div class="modal-footer">
</div>
</div>
</div>
</div>
<script src="https://unpkg.com/mermaid@8.8.0/dist/mermaid.min.js"></script><script>mermaid.initialize({});</script></body>
</html>
<!--
MkDocs version : 1.1.2
Build Date UTC : 2021-08-29 12:17:10.091521+00:00
-->
