<!DOCTYPE html>

<html lang="en">
<head>
<meta charset="utf-8"/>
<meta content="IE=edge" http-equiv="X-UA-Compatible"/>
<meta content="width=device-width, initial-scale=1.0" name="viewport"/>
<link href="../img/favicon.ico" rel="shortcut icon"/>
<title>Notes - SCALLOP-INF</title>
<link href="../css/bootstrap.min.css" rel="stylesheet"/>
<link href="../css/font-awesome.min.css" rel="stylesheet"/>
<link href="../css/base.css" rel="stylesheet"/>
<link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css" rel="stylesheet"/>
<script defer="" src="../js/jquery-1.10.2.min.js"></script>
<script defer="" src="../js/bootstrap.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/Bash.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/HTML.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/Java.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/JavaScript.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/Perl.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/Python.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/R.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/rust.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/yaml.min.js"></script>
<script>hljs.initHighlightingOnLoad();</script>
</head>
<body>
<div class="navbar fixed-top navbar-expand-lg navbar-dark bg-dark">
<div class="container">
<a class="navbar-brand" href="..">SCALLOP-INF</a>
<!-- Expander button -->
<button class="navbar-toggler" data-target="#navbar-collapse" data-toggle="collapse" type="button">
<span class="navbar-toggler-icon"></span>
</button>
<!-- Expanded navigation -->
<div class="navbar-collapse collapse" id="navbar-collapse">
<!-- Main navigation -->
<ul class="nav navbar-nav">
<li class="navitem">
<a class="nav-link" href=".."></a>
</li>
<li class="navitem">
<a class="nav-link" href="../doc/SCALLOP_INF1_analysis_plan/">Analysis plan</a>
</li>
<li class="dropdown active">
<a class="nav-link dropdown-toggle" data-toggle="dropdown" href="#">Implementation <b class="caret"></b></a>
<ul class="dropdown-menu">
<li>
<a class="dropdown-item" href="../tryggve/">Tryggve</a>
</li>
<li>
<a class="dropdown-item" href="../cardio/">Cardio</a>
</li>
<li>
<a class="dropdown-item" href="../csd3/">CSD3</a>
</li>
<li>
<a class="dropdown-item active" href="./">Notes</a>
</li>
</ul>
</li>
</ul>
<ul class="nav navbar-nav ml-auto">
<li class="nav-item">
<a class="nav-link" data-target="#mkdocs_search_modal" data-toggle="modal" href="#">
<i class="fa fa-search"></i> Search
                            </a>
</li>
<li class="nav-item">
<a class="nav-link" href="../csd3/" rel="prev">
<i class="fa fa-arrow-left"></i> Previous
                                </a>
</li>
<li class="nav-item">
<a class="nav-link disabled" rel="next">
                                    Next <i class="fa fa-arrow-right"></i>
</a>
</li>
</ul>
</div>
</div>
</div>
<div class="container">
<div class="row">
<div class="col-md-3"><div class="navbar-light navbar-expand-md bs-sidebar hidden-print affix" role="complementary">
<div class="navbar-header">
<button class="navbar-toggler collapsed" data-target="#toc-collapse" data-toggle="collapse" title="Table of Contents" type="button">
<span class="fa fa-angle-down"></span>
</button>
</div>
<div class="navbar-collapse collapse card bg-secondary" id="toc-collapse">
<ul class="nav flex-column">
<li class="nav-item" data-level="1"><a class="nav-link" href="#notes">Notes</a>
<ul class="nav flex-column">
<li class="nav-item" data-level="2"><a class="nav-link" href="#jointconditional-analysis-and-fine-mapping">Joint/conditional analysis and fine-mapping</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#mendelian-randomization-mr-analysis">Mendelian Randomization (MR) analysis</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#variant-annotation">Variant annotation</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#colocalization-analysis">Colocalization analysis</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#generation-of-summary-statistics">Generation of summary statistics</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#mathematical-expressions">Mathematical expressions</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#nested-pgs-model">Nested PGS model</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#auxiliary-files">Auxiliary files</a>
<ul class="nav flex-column">
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div></div>
<div class="col-md-9" role="main">
<h1 id="notes">Notes</h1>
<h2 id="jointconditional-analysis-and-fine-mapping">Joint/conditional analysis and fine-mapping</h2>
<p>The INTERVAL data is used as reference panel. The logic of this specific directory is a simple solution of the dilemma that the reference data, possibly like others, uses reference sequence ID (rsid) whenever possible. However, during meta-analysis the practice of using rsids is undesirable so SNPID, i.e., chr:pos_A1_A2, (A1&lt;=A2) is necessary.</p>
<p>After a rather long and laborious process involving many software, it turned out a simple way out is to obtain sentinels using SNPID but return to rsids at this stage and forward. The implementation here reflects this. The file INTERVAL.rsid contains SNPID-rsid mapping and could be generated from programs such as <code>qctool/bgenix/plink</code>. A related file is <code>snp_pos</code> containing rsid -- chr:pos mappings; for instance this could be built from dbSNP.</p>
<p>A note on regions is ready. It is attractive to use the last genomic region from iterative merging for analysis and perhaps a flanking version. This is more appropriate than genome-wide hard and fast 10MB windows or approximately independent LD blocks. For the latter, we found that the boundaries from the distributed 1000Genomes project were often inappropriate and one may not attempt to compute them for specific reference panel. Nevertheless, the iterative procedure actually just does empirically. Again the HLA region is condensed.</p>
<p>The last point regards software <code>finemap</code>, which uses summary statistics associated with the reference panel rather than that from meta-analysis.</p>
<table>
<thead>
<tr>
<th>File</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>NLRP2.sh</td>
<td>the exclusion list</td>
</tr>
<tr>
<td>cs.sh</td>
<td>Credible sets</td>
</tr>
<tr>
<td>ma.sh</td>
<td>INF1 sumstats</td>
</tr>
<tr>
<td>INTERVAL-ma.sh</td>
<td>INTERVAL sumstats</td>
</tr>
<tr>
<td>prune.sh</td>
<td>pruning</td>
</tr>
<tr>
<td>slct.sh</td>
<td>GCTA --cojo-slct analysis</td>
</tr>
<tr>
<td>finemap.sh</td>
<td><code>finemap</code> analysis</td>
</tr>
<tr>
<td>jam.sh</td>
<td><code>JAM</code> analysis</td>
</tr>
<tr>
<td>st.sh</td>
<td>batch command file</td>
</tr>
<tr>
<td>cs/, finemap/, jam/, prune/, work/</td>
<td>working directories</td>
</tr>
</tbody>
</table>
<p><strong>Steps</strong></p>
<p><code>st.sh</code> conceptually executes the following elements,</p>
<div class="mermaid">graph TB
sumstats("sumstats from METAL") --&gt; |"variants from prune.sh"| slct.sh
utils.sh("INTERVAL sumstats from standardized residuals ~ genotypes") --&gt; |"variants from prune.sh"| cs.sh
utils.sh --&gt; |"internal prune" | jam.sh
utils.sh  --&gt; |"--n-causal-snps k"| finemap.sh
</div>
<p>Note that the <code>GCTA</code> .ma, jma.cojo, .ldr.cojo become -rsid.ma, -rsid.jma.cojo, -rsid.ldr.cojo, respectively; the same are true for files related to <code>finemap</code>.</p>
<h2 id="mendelian-randomization-mr-analysis">Mendelian Randomization (MR) analysis</h2>
<p>There has been a major overhaul by refocusing on cis-pQTLs using GSMR involving both immune-mediated traits and COVID outcomes. These have been implemented in <code>mr.sh</code> and <code>mr.sb</code>. To get around the mishap with rsids we switched back to SNPids; we found the options <code>--clump-r2 0.1 --gsmr-snp-min 10</code> make more sense for instance with IL-12B it would retain the sentinel and also the significant MR results on inflammatory bowel disease. Earlier we coded specifically for HGI analysis, in particular aiming for four test statistics and better quality graphics.</p>
<p>It turned to be more efficient in cis-pQTL analyses to extract the relevant variants involving in the analysis, rather than using whole-genome reference files. This led to ref/, prot/, /trait directories for reference, protein and trait data, respectively.</p>
<p>The GSMR implementation started with CAD/FEV1 (now with <code>gsmr.sh</code>, <code>gsmr.sb</code>, <code>gsmr.R</code>) and then HGI as <code>gsmr_HGI.*</code> and pQTLMR/MR (<code>HGI.sh</code>, <code>HGI.R</code>), and +/- 1MB (<code>rs635634.sh</code>).</p>
<p>Earlier on, a pQTL-based MR analysis is furnished with <code>pqtlMR.sh</code>.</p>
<h2 id="variant-annotation">Variant annotation</h2>
<p>This is exemplified with <code>vep.sh</code>. GARFIELD analysis is furnsihed with <code>garfield.sh</code>. Specific and collective enrichment can be found in <code>magma.sh</code> and <code>rGREAT.R</code>.</p>
<h2 id="colocalization-analysis">Colocalization analysis</h2>
<p>eQTL lookup is done with PhenoScanner via <code>eQTL.R</code>, <code>eQTLGen.sh</code> and GTEx via <code>GTEx-ps.R</code>, <code>GTEx-ps.sh</code> and <code>GTEx-ps-jma.sh</code>.</p>
<p>Credible set lookup from GTEx v8 is furnished with <code>GTEx-cs.sh</code> and <code>GTEx-cs-jma.sh</code>.</p>
<p>Eearlier experiments were done with <code>coloc.sb</code> (clumsy) and <code>coloc.R</code> (with pQTLtools) as well as <code>fastenloc.sb</code>.</p>
<p>Multitrait counterparts were also done with HyPrColoc as in <code>rs12075.sh</code> and <code>LTBR.sh</code>.</p>
<h2 id="generation-of-summary-statistics">Generation of summary statistics</h2>
<p>This is implemented with <code>gwasvcf.sh</code> and <code>gwas2vcf.sb</code> which includes some operations on VCF files. However, we found there is loss of information if enforcing use of RSid.</p>
<h2 id="mathematical-expressions">Mathematical expressions</h2>
<ol>
<li>
<p><code>gap::get_pve_se</code>: The proportion of variants explained (PVE) by <span class="arithmatex">\(T\)</span> pQTLs for a protein from meta-analytic statistics is</p>
<p>PVE=<span class="arithmatex">\(\sum_{i=1}^T{\frac{\chi_i^2}{N_i-2+\chi_i^2}}\)</span></p>
<p>with standard error</p>
<p><span class="arithmatex">\(SE_{PVE}=\sum_{i=1}^T{\frac{1}{N_i-1}}\)</span></p>
<p>This is according to t-statistic; an alternative form also uses regression theory. Let <span class="arithmatex">\(\mbox{x} = SNP\ dosage\)</span>. From <span class="arithmatex">\(\mbox{y}=a + b\mbox{x} + e\)</span> we have <span class="arithmatex">\(Var(\mbox{y}) = b^2Var(\mbox{x}) + Var(e)\)</span>. Note that <span class="arithmatex">\(Var(\mbox{x})=2f(1-f)\)</span>, <span class="arithmatex">\(f=MAF\)</span> or <span class="arithmatex">\(1-MAF\)</span> by symmetry. In a linear regression <span class="arithmatex">\(Var(b)=Var(e)/S_\mbox{xx}\)</span>, we have <span class="arithmatex">\(Var(e) = Var(b)S_\mbox{xx} = N Var(b) Var(\mbox{x})\)</span> and therefore <span class="arithmatex">\(PVE = Var(b\mbox{x}) = b^22f(1-f)/(b^22f(1-f)+2NVar(b)f(1-f)}\)</span>. In fact, let <span class="arithmatex">\(z=b/SE(b)\)</span>, <span class="arithmatex">\(PVE=\mbox{z}^2/(\mbox{z}^2+N)\)</span>.</p>
<p>The standard errors of both forms can be obtained via variance for a ratio (R/S). First, we state some established results:</p>
<div class="arithmatex">\[
\begin{align}
E(R/S) \equiv E(f(R,S))
       \approx \frac{\mu_R}{\mu_S}-\frac{\Cov(R,S)}{(\mu_S)^2}+\frac{\Var(S)\mu_R}{(\mu_S)^3} \hspace{100cm}
\end{align}
\]</div>
<p>and variance</p>
<div class="arithmatex">\[
\begin{align}
\Var(R/S) &amp;\approx&amp; \frac{1}{(\mu_S)^2}\Var(R)
            +2\frac{-\mu_R}{(\mu_S)^3}\Cov(R,S)+\frac{(\mu_R)^2}{(\mu_S)^4}\Var(S)  \hspace{100cm} \\
         &amp;=&amp; \frac{(\mu_R)^2}{(\mu_S)^2} \left[
             \frac{\Var(R)}{(\mu_R)^2}-2\frac{\Cov(R,S)}{\mu_R \; \mu_S}
            +\frac{\Var(S)}{(\mu_S)^2} \right] \\
         &amp;=&amp; \frac{(\mu_R)^2}{(\mu_S)^2} \left[
             \frac{\sigma_R^2}{(\mu_R)^2} -2\frac{\Cov(R,S)}{\mu_R\;\mu_S}
            +\frac{\sigma_S^2}{(\mu_S)^2} \right]
\end{align}
\]</div>
<p>In addition, <span class="arithmatex">\(E(\chi_1^2)=1\)</span> and <span class="arithmatex">\(Var(\chi_1^2)=2\)</span>, the variances are <span class="arithmatex">\(\frac{2}{(N-1)^2}\)</span> and <span class="arithmatex">\(\frac{2}{(N+1)^2}\)</span>, respectively by ignoring a high-order term.</p>
</li>
<li>
<p>The prior probability for <span class="arithmatex">\(i\)</span> out of <span class="arithmatex">\(m\)</span> variants is causal, is obtained as a binomial probability</p>
<p><span class="arithmatex">\(p(i)={m\choose{i}}{\left(\frac{i}{m}\right)^i\left(1-\frac{i}{m}\right)^{m-i}}\)</span></p>
<p>The posterior number of causal signals in the genomic region is obtained as the expectation <span class="arithmatex">\(\sum_i p(i|\mbox{data}) \times i\)</span>. The search of total number of configurations</p>
<p><span class="arithmatex">\(\sum_{i=1}^k{m\choose{i}}\)</span></p>
</li>
<li>
<p><code>gap::get_b_se</code>: To recover the effect size (<span class="arithmatex">\(b\)</span>) and its standard error (<span class="arithmatex">\(se\)</span>) from allele frequency (<span class="arithmatex">\(f\)</span>), sample size (<span class="arithmatex">\(n\)</span>) and z-statistic (<span class="arithmatex">\(z\)</span>) as in cis eQTLGen data, we have</p>
<div class="arithmatex">\[
\begin{align}
b &amp; = z/d \hspace{100cm} \\ 
se &amp; = 1/d
\end{align}
\]</div>
<p>where <span class="arithmatex">\(d = \sqrt{2f(1-f)(n+z^2)}\)</span>.</p>
</li>
<li>
<p>The GREAT Binomial test is</p>
<p><span class="arithmatex">\(\sum_{i=k_\pi}^n{n\choose{i}}p_\pi^i(1-p_\pi)^{n-i}\)</span></p>
</li>
<li>
<p>A proteome-wide analogy <span class="arithmatex">\(z_{PWAS}\)</span> is made to (transcriptome)wide association statistic originally proposed for gene expression data. For a given Protein of interest <strong>P</strong> for which GWAS summary statistics <span class="arithmatex">\(z_P\)</span> is available, the corresponding Wald statistic for PWAS is defined such that</p>
<p><span class="arithmatex">\(z_{PWAS} = \frac{w^T_{P}z_P}{\sqrt{w^T_{P}Vw_{P}}}\)</span></p>
<p>where <span class="arithmatex">\(w_{P}\)</span> is a weight associated with protein abundance level and <strong>V</strong> covariance matrix for <span class="arithmatex">\(z_P\)</span>, respectively.</p>
</li>
</ol>
<h2 id="nested-pgs-model">Nested PGS model</h2>
<p align="center"><img src="grViz.png"/></p>
<p>Without loss of generality, we have genotype data G, which link with proteins P1, P2, P3 as predictors for outcome y.</p>
<p>The figure is obtained with</p>
<pre><code class="language-bash">dot -Tpng grViz.gv -ogrViz.png
</code></pre>
<p>which also be  obtained from RStudio for somewhat larger size.</p>
<hr/>
<h2 id="auxiliary-files">Auxiliary files</h2>
<table>
<thead>
<tr>
<th>File</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>chembl.sh</td>
<td>toy CHEMBL queries</td>
</tr>
<tr>
<td>CRP.sh</td>
<td>an inflammation score</td>
</tr>
<tr>
<td>cvd1.sh</td>
<td>SCALLOP-CVDI supplementary tables</td>
</tr>
<tr>
<td>circos.*</td>
<td>circos/circlize scripts</td>
</tr>
<tr>
<td>efo.R</td>
<td>experimental factor ontology</td>
</tr>
<tr>
<td>epigraphdb-pleiotropy.R</td>
<td>horizontal and vertical pleiotropy</td>
</tr>
<tr>
<td>epigraphdb-ppi.R</td>
<td>PPI using EpiGraphDB</td>
</tr>
<tr>
<td>fusion_twas.sb</td>
<td>FUSION TWAS experiment</td>
</tr>
<tr>
<td>gdigt.R</td>
<td>GDI and gene-trials</td>
</tr>
<tr>
<td>GREP.sh</td>
<td>GREP script</td>
</tr>
<tr>
<td>h2pve.R</td>
<td>h2/pve contrast</td>
</tr>
<tr>
<td>inbio-discover.R</td>
<td>gene lists</td>
</tr>
<tr>
<td>neale.sh</td>
<td>signal selection for Neale UKB data (HbA1c)</td>
</tr>
<tr>
<td>OpenGWAS.sh</td>
<td>OpenGWAS download</td>
</tr>
<tr>
<td>pheweb.Rmd</td>
<td>Data extraction from pheweb.jp</td>
</tr>
<tr>
<td>pqtlGWAS.R</td>
<td>pQTL-GWAS lookup</td>
</tr>
<tr>
<td>pQTL.R</td>
<td>pQTL lookup</td>
</tr>
<tr>
<td>qqman.sb</td>
<td>Q-Q/Manhattan plots under CSD3</td>
</tr>
<tr>
<td>rentrez.sh</td>
<td>reuse of rentrez</td>
</tr>
<tr>
<td>Somascan-Olink.*</td>
<td>Somascan/Olink overlap</td>
</tr>
<tr>
<td>stringdb.sh</td>
<td>STRINGdb</td>
</tr>
<tr>
<td>tables.R</td>
<td>code to create Excel Tables</td>
</tr>
<tr>
<td>uniprot.R</td>
<td>UniProt IDs to others</td>
</tr>
<tr>
<td>utils.sh</td>
<td>utilties</td>
</tr>
<tr>
<td>wgcna.sh</td>
<td>experiment on modules</td>
</tr>
</tbody>
</table>
<h3 id="stacked-association-and-related-analysis">Stacked association and related analysis</h3>
<ol>
<li>rs12075.sh</li>
<li>MS.sh (+cojo, two-sample MRs)+LTBR.sh (stack_assoc_plot + HyPrColoc/PWCoCo)</li>
<li>IL.18-rs385076.sh</li>
<li>OPG-TRANCE.sh</li>
</ol>
<p>and <code>MS.sh</code> contains a routine to generate tabix-indexed blood cell traits.</p>
<h3 id="uniprot-ids">UniProt IDs</h3>
<p>The list of proteins on inflammation is within one of the <a href="https://www.olink.com/products/">Olink</a> panels (as in <a href="../doc/Olink%20validation%20data%20all%20panels.xlsx">Olink validation data all panels.xlsx</a>) each containing 92 proteins. Information contained in these panels can be retrieved into R via <a href="../doc/Olink.R">Olink.R</a>, which also attempts to compromise earlier version and annotations. Nevertheless O43508 is replaced with Q4ACW9 for TWEAK.</p>
<p>BDNF has recently been removed from the assay and replaced with CD8A, <a href="https://www.olink.com/bdnf-info/">https://www.olink.com/bdnf-info/</a>, and there are also changes on TNF and IFN.gamma, <a href="https://www.olink.com/inflammation-upgrade/">https://www.olink.com/inflammation-upgrade/</a>.</p>
<p>A <a href="https://www.uniprot.org/">UniProt</a> ID may be associated with multiple chromosomes, e.g., Q6IEY1 with chromosomes 1 and 5. While <a href="../doc/inf1.csv">inf1.csv</a> 
edits Q4ACW9, <a href="../doc/inf2.csv">inf2.csv</a> is inline with UCSC with respect to P12034 and P30203.</p>
<p>The use of uniprot IDs is noted in two aspects,</p>
<ol>
<li>
<p>The protein list in <a href="../doc/inf1.csv">inf1.csv</a> notes both O43508 and Q4ACW9.</p>
</li>
<li>
<p>Q8NF90 and Q8WWJ7 were not listed at the UCSC, their availability on UniProt seem to be for backward compatibility as on query they 
point to P12034 and P30203 (Q8WWJ7_HUMAN should have been CD6_HUMAN). <a href="../doc/hgTables.tsv">hgTables.tsv</a> is based on UCSC, checked over
UniProt IDs as follows,</p>
</li>
</ol>
<pre><code class="language-bash">grep inf1 doc/olink.prot.list.txt | \
sed 's/inf1_//g;s/___/\t/g' &gt; inf1.list
join -t$'\t' -12 -24 \
     &lt;(sort -k2,2 inf1.list) \
     &lt;(awk '{split($4,f,"-"); $4=f[1]; if(!index($1,"_")) print}' OFS='\t' doc/hgTables.tsv | sort -k4,4) \
     &gt; 12
# 90 lines
wc -l 12
# Q8NF90 (FGF.5), Q8WWJ7 (CD6) are missing
join -v2 -22 12 &lt;(sort -k2,2 inf1.list)
rm 12
</code></pre>
<p>Likewise, <a href="../doc/olink.inf.panel.annot.tsv">olink.inf.panel.annot.tsv</a> from <a href="../doc/olink.annotation.R">olink.annotation.R</a> also has the following two entries</p>
<table>
<thead>
<tr>
<th>"target"</th>
<th>"target.short"</th>
<th>"uniprot"</th>
<th>"panel"</th>
<th>"prot.on.multiple.panel"</th>
<th>"panels.with.prot"</th>
<th>"hgnc_symbol"</th>
<th>...</th>
</tr>
</thead>
<tbody>
<tr>
<td>"Fibroblast growth factor 5 (FGF-5)"</td>
<td>"FGF-5"</td>
<td>"Q8NF90"</td>
<td>"inf"</td>
<td>FALSE</td>
<td>NA</td>
<td>NA</td>
<td></td>
</tr>
<tr>
<td>"T-cell surface glycoprotein CD6 isoform (CD6)"</td>
<td>"CD6"</td>
<td>"Q8WWJ7"</td>
<td>"inf"</td>
<td>FALSE</td>
<td>NA</td>
<td>NA</td>
<td></td>
</tr>
</tbody>
</table>
<p>whose hgnc_symbol can be amended.</p>
<p><img align="right" height="300" src="../doc/Olink-SomaLogic-Venn-diagram.png" width="300"/></p>
<p>The overlap with SomaLogic panel is characterised with <a href="../doc/Olink.R">Olink.R</a> which also gives a Venn diagram.</p>
<p><em>Date last changed:</em> <strong>4/11/2021</strong></p></div>
</div>
</div>
<footer class="col-md-12">
<hr/>
<p>Documentation built with <a href="https://www.mkdocs.org/">MkDocs</a>.</p>
</footer>
<script>
            var base_url = "..",
                shortcuts = {"help": 191, "next": 78, "previous": 80, "search": 83};
        </script>
<script defer="" src="../js/base.js"></script>
<script defer="" src="../javascripts/config.js"></script>
<script defer="" src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script defer="" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
<script defer="" src="../search/main.js"></script>
<div aria-hidden="true" aria-labelledby="searchModalLabel" class="modal" id="mkdocs_search_modal" role="dialog" tabindex="-1">
<div class="modal-dialog modal-lg">
<div class="modal-content">
<div class="modal-header">
<h4 class="modal-title" id="searchModalLabel">Search</h4>
<button class="close" data-dismiss="modal" type="button"><span aria-hidden="true">×</span><span class="sr-only">Close</span></button>
</div>
<div class="modal-body">
<p>
                    From here you can search these documents. Enter
                    your search terms below.
                </p>
<form>
<div class="form-group">
<input class="form-control" id="mkdocs-search-query" placeholder="Search..." title="Type search term here" type="search"/>
</div>
</form>
<div id="mkdocs-search-results"></div>
</div>
<div class="modal-footer">
</div>
</div>
</div>
</div><div aria-hidden="true" aria-labelledby="keyboardModalLabel" class="modal" id="mkdocs_keyboard_modal" role="dialog" tabindex="-1">
<div class="modal-dialog">
<div class="modal-content">
<div class="modal-header">
<h4 class="modal-title" id="keyboardModalLabel">Keyboard Shortcuts</h4>
<button class="close" data-dismiss="modal" type="button"><span aria-hidden="true">×</span><span class="sr-only">Close</span></button>
</div>
<div class="modal-body">
<table class="table">
<thead>
<tr>
<th style="width: 20%;">Keys</th>
<th>Action</th>
</tr>
</thead>
<tbody>
<tr>
<td class="help shortcut"><kbd>?</kbd></td>
<td>Open this help</td>
</tr>
<tr>
<td class="next shortcut"><kbd>n</kbd></td>
<td>Next page</td>
</tr>
<tr>
<td class="prev shortcut"><kbd>p</kbd></td>
<td>Previous page</td>
</tr>
<tr>
<td class="search shortcut"><kbd>s</kbd></td>
<td>Search</td>
</tr>
</tbody>
</table>
</div>
<div class="modal-footer">
</div>
</div>
</div>
</div>
<script src="https://unpkg.com/mermaid@8.8.0/dist/mermaid.min.js"></script><script>mermaid.initialize({});</script></body>
</html>
