<!DOCTYPE html>

<html lang="en">
<head>
<meta charset="utf-8"/>
<meta content="IE=edge" http-equiv="X-UA-Compatible"/>
<meta content="width=device-width, initial-scale=1.0" name="viewport"/>
<link href="../img/favicon.ico" rel="shortcut icon"/>
<title>CSD3 - SCALLOP-INF</title>
<link href="../css/bootstrap.min.css" rel="stylesheet"/>
<link href="../css/font-awesome.min.css" rel="stylesheet"/>
<link href="../css/base.css" rel="stylesheet"/>
<link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css" rel="stylesheet"/>
<script defer="" src="../js/jquery-1.10.2.min.js"></script>
<script defer="" src="../js/bootstrap.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/Bash.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/HTML.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/Java.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/JavaScript.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/Perl.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/Python.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/R.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/rust.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/yaml.min.js"></script>
<script>hljs.initHighlightingOnLoad();</script>
</head>
<body>
<div class="navbar fixed-top navbar-expand-lg navbar-dark bg-dark">
<div class="container">
<a class="navbar-brand" href="..">SCALLOP-INF</a>
<!-- Expander button -->
<button class="navbar-toggler" data-target="#navbar-collapse" data-toggle="collapse" type="button">
<span class="navbar-toggler-icon"></span>
</button>
<!-- Expanded navigation -->
<div class="navbar-collapse collapse" id="navbar-collapse">
<!-- Main navigation -->
<ul class="nav navbar-nav">
<li class="navitem">
<a class="nav-link" href=".."></a>
</li>
<li class="navitem">
<a class="nav-link" href="../doc/SCALLOP_INF1_analysis_plan/">Analysis plan</a>
</li>
<li class="dropdown active">
<a class="nav-link dropdown-toggle" data-toggle="dropdown" href="#">Implementation <b class="caret"></b></a>
<ul class="dropdown-menu">
<li>
<a class="dropdown-item" href="../tryggve/">Tryggve</a>
</li>
<li>
<a class="dropdown-item" href="../cardio/">Cardio</a>
</li>
<li>
<a class="dropdown-item active" href="./">CSD3</a>
</li>
<li>
<a class="dropdown-item" href="../rsid/">Notes</a>
</li>
</ul>
</li>
</ul>
<ul class="nav navbar-nav ml-auto">
<li class="nav-item">
<a class="nav-link" data-target="#mkdocs_search_modal" data-toggle="modal" href="#">
<i class="fa fa-search"></i> Search
                            </a>
</li>
<li class="nav-item">
<a class="nav-link" href="../cardio/" rel="prev">
<i class="fa fa-arrow-left"></i> Previous
                                </a>
</li>
<li class="nav-item">
<a class="nav-link" href="../rsid/" rel="next">
                                    Next <i class="fa fa-arrow-right"></i>
</a>
</li>
</ul>
</div>
</div>
</div>
<div class="container">
<div class="row">
<div class="col-md-3"><div class="navbar-light navbar-expand-md bs-sidebar hidden-print affix" role="complementary">
<div class="navbar-header">
<button class="navbar-toggler collapsed" data-target="#toc-collapse" data-toggle="collapse" title="Table of Contents" type="button">
<span class="fa fa-angle-down"></span>
</button>
</div>
<div class="navbar-collapse collapse card bg-secondary" id="toc-collapse">
<ul class="nav flex-column">
<li class="nav-item" data-level="1"><a class="nav-link" href="#csd3">CSD3</a>
<ul class="nav flex-column">
<li class="nav-item" data-level="2"><a class="nav-link" href="#3d-scatter-plot">3d scatter plot</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#annotation">Annotation</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#known-variants">Known variants</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#ld-reference-panels">LD reference panels</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#snpid-rsid-mappings-conditional-analysis-and-finemapping">SNPID-rsid mappings, conditional analysis and finemapping</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#low-abundance-check">Low abundance check</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#pgs">PGS</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#phenoscanner">PhenoScanner</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#qq-and-manhattan-plots">QQ and Manhattan plots</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#sentinel-identification">Sentinel identification</a>
<ul class="nav flex-column">
</ul>
</li>
<li class="nav-item" data-level="2"><a class="nav-link" href="#trans-pqtl-hotspots-and-proteins-as-polygenic-targets">Trans-pQTL hotspots and proteins as polygenic targets</a>
<ul class="nav flex-column">
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div></div>
<div class="col-md-9" role="main">
<h1 id="csd3">CSD3</h1>
<p>CSD3 is the HPC Linux clusters used at the University of Cambridge, with which this document is associated.</p>
<p><strong>NOTE</strong>: <a href="merge.sh">merge.sh</a> and <a href="INTERVAL-fm.ini">INTERVAL-fm.ini</a> now work directly with the original INTERVAL bgen files. The alphabetical list is as follows.</p>
<h2 id="3d-scatter-plot">3d scatter plot</h2>
<p>Furnished with <a href="js.R">js.R</a>, the JSON output could be used as data in the Supplementary Figure 1 of Sun et al. (2018).</p>
<h2 id="annotation">Annotation</h2>
<p><a href="annotate.sh">annotate.sh</a> involves <strong>ANNOVAR</strong>, <strong>PolyPhen 2</strong>, <strong>VEP</strong> and some R packages.</p>
<h2 id="known-variants">Known variants</h2>
<p><a href="lookup.sh">lookup.sh</a>, which uses PhenoScanner output from <a href="pqtl.R">pqtl.R</a>, can be used to check if pQTLs were known.</p>
<p>File <a href="cvd1.sh">cvd1.sh</a> gives sentinels overlapped with CVD I as well as LD information corresponding to each protein.</p>
<h2 id="ld-reference-panels">LD reference panels</h2>
<p>The .sh versions below extract data from INTERVAL and ukb, each calling a .sb to generate <code>binary_ped</code> with SNPIDs (bgen/<em>map)
* by </em><em>qctool</em><em> -- it also creates <code>bgen</code> with SNPIDs (ukb/nodup) to avoid hard-called genotypes.
* by </em><em>PLINK</em>* -- it has a <code>_snpid</code> tag.</p>
<p>In both cases duplicates (bgen/*rmdup.list) are excluded.</p>
<ol>
<li>The utility <a href="INTERVAL.sh">INTERVAL.sh</a> extracts data from  </li>
</ol>
<pre><code>/DO-NOT-MODIFY-SCRATCH/bp406/data_sets/interval_subset_olink/genotype_files/unrelated_4994_pihat_0.1875_autosomal_typed_only
/DO-NOT-MODIFY-SCRATCH/bp406/data_sets/interval_subset_olink/genotype_files/unrelated_4994_pihat_0.1875_autosomal_imputed_info_0.4_phwe_1e-4_filtered/per_chr
</code></pre>
<p>on Cardio into region-specific data in <code>bgen</code> format according to work/INF1.merge.</p>
<ol>
<li>The utility <a href="ukb.sh">ukb.sh</a> extracts data from ukb_imp_chr[x-xx]_v3.bgen as in</li>
</ol>
<pre><code>/DO-NOT-MODIFY-SCRATCH/uk_biobank/500k/imputed_v3
/DO-NOT-MODIFY-SCRATCH/curated_genetic_data/uk_biobank/reference_files/full_release
</code></pre>
<p>on Cardio into region-specific data in <code>bgen</code> format (ukb/bgen) according to work/INF1.merge.</p>
<h2 id="snpid-rsid-mappings-conditional-analysis-and-finemapping">SNPID-rsid mappings, conditional analysis and finemapping</h2>
<p>This is furnished with <a href="snpid_rsid.sb">snpid_rsid.sb</a>, whose results will be attached to <strong>GCTA</strong>/<strong>finemap</strong> results via <code>finemap.R</code>, <code>slct.R</code> and <code>jam.R</code>.
After that, it is ready to use script <code>slct.sb</code> followed by <code>slct.sh</code>. Optionally, the results are fed into <code>finemap.sb</code> via <code>--n-causal-snps</code>.</p>
<ul>
<li><a href="finemap.sb">finemap.sb</a> uses the unpruned version.</li>
<li><a href="fm.sb">fm.sb</a> and <a href="INTERVAL-fm.sb">INTERVAL-fm.sb</a> use pruned variants to compromise <code>JAM</code>. The <code>.z</code> file is also appropriate for both <code>finemap</code> and <code>JAM</code>.</li>
<li><a href="INTERVAL-fm.sb">INTERVAL-fm.sh</a> and <a href="INTERVAL-fm.ini">INTERVAL-fm.ini</a> works on INTERVAL data.</li>
</ul>
<pre><code class="language-bash"># ldstore v1.1

export pr=IL.6-chr1:154426970_A_C

ldstore --bcor ${pr} --bgen ${pr}.bgen
ldstore --bcor ${pr} --merge 1
ldstore --bcor ${pr} --matrix ${pr}.ld

# ldstore 2 (and finemap-1.4)

ldstore --in-files ${pr}.master2 --write-bcor --write-bdose

finemap --sss --in-files ${pr}.master2 --n-causal-snps 10
</code></pre>
<p>where ${pr}.master2 contains two lines,</p>
<pre><code>z;bgen;bgi;bcor;bdose;snp;config;cred;log;n_samples
IL.6-chr1:154426970_A_C.z;IL.6-chr1:154426970_A_C.bgen;IL.6-chr1:154426970_A_C.bgi;IL.6-chr1:154426970_A_C.bcor;IL.6-chr1:154426970_A_C.bdose;IL.6-chr1:154426970_A_C.snp;IL.6-chr1:154426970_A_C.config;IL.6-chr1:154426970_A_C.cred;IL.6-chr1:154426970_A_C.log;4994
</code></pre>
<p>The environmental variable <code>LDREF</code> provides an option to use either INTERVAL or UKB data.</p>
<h2 id="low-abundance-check">Low abundance check</h2>
<p>This is down with <a href="qc.sh">qc.sh</a>.</p>
<h2 id="pgs">PGS</h2>
<p>These are <a href="prsice.sh">prsice.sh</a> and <a href="pgs.sh">pgs.sh</a> for the CRP example.</p>
<h2 id="phenoscanner">PhenoScanner</h2>
<p>This is furnished with <a href="ps.sh">ps.sh</a>.</p>
<h2 id="qq-and-manhattan-plots">QQ and Manhattan plots</h2>
<p>The version for meta-analysis was part of <code>analysis.sh</code> at tryggve/. <a href="qqman.sh">qqman.sh</a> calls turboqq and turboman by Bram Prins.</p>
<h2 id="sentinel-identification">Sentinel identification</h2>
<p>This is now furnished with <a href="merge.sh">merge.sh</a>.</p>
<p>Clumping by fixed distance is superseded with its failure to handle long LD regions.</p>
<p>There are scripts for heritability analysis and proportion of variance explained.</p>
<h2 id="trans-pqtl-hotspots-and-proteins-as-polygenic-targets">Trans-pQTL hotspots and proteins as polygenic targets</h2>
<p>This is illustrated with circos plots and by default this works on <code>work/INF1.merge.cis.vs.trans</code> and requires <a href="glist-hg19">glist-hg19</a>.
Because circos plots are gene-centric, in both cases, the protein-coding gene is handled in mind.</p>
<p>Try</p>
<pre><code class="language-bash">./hotspot.sh chr1:159175354_A_G
./hotspot.sh chr12:111884608_C_T
</code></pre>
<p>giving results linking <em>ACKR1</em> and <em>SH2B3</em>, respectively, while</p>
<pre><code class="language-bash">./polygene.sh IL12B
./polygene.sh TNFSF10
</code></pre>
<p>linking IL.12B and TRAIL, respectively.</p>
<p>To generate all possible plots, wo do</p>
<pre><code class="language-bash">for h in $(cut -f6 work/INF1.merge | sed '1d' | sort -k1,1 | uniq); do echo $h; hotspot.sh $h; done
for g in $(cat work/INF1.merge.gene); do echo $g; polygene.sh $g; done
</code></pre>
<p><em>Date laste changed:</em> <strong>7/10/2021</strong></p></div>
</div>
</div>
<footer class="col-md-12">
<hr/>
<p>Documentation built with <a href="https://www.mkdocs.org/">MkDocs</a>.</p>
</footer>
<script>
            var base_url = "..",
                shortcuts = {"help": 191, "next": 78, "previous": 80, "search": 83};
        </script>
<script defer="" src="../js/base.js"></script>
<script defer="" src="../javascripts/config.js"></script>
<script defer="" src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script defer="" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
<script defer="" src="../search/main.js"></script>
<div aria-hidden="true" aria-labelledby="searchModalLabel" class="modal" id="mkdocs_search_modal" role="dialog" tabindex="-1">
<div class="modal-dialog modal-lg">
<div class="modal-content">
<div class="modal-header">
<h4 class="modal-title" id="searchModalLabel">Search</h4>
<button class="close" data-dismiss="modal" type="button"><span aria-hidden="true">×</span><span class="sr-only">Close</span></button>
</div>
<div class="modal-body">
<p>
                    From here you can search these documents. Enter
                    your search terms below.
                </p>
<form>
<div class="form-group">
<input class="form-control" id="mkdocs-search-query" placeholder="Search..." title="Type search term here" type="search"/>
</div>
</form>
<div id="mkdocs-search-results"></div>
</div>
<div class="modal-footer">
</div>
</div>
</div>
</div><div aria-hidden="true" aria-labelledby="keyboardModalLabel" class="modal" id="mkdocs_keyboard_modal" role="dialog" tabindex="-1">
<div class="modal-dialog">
<div class="modal-content">
<div class="modal-header">
<h4 class="modal-title" id="keyboardModalLabel">Keyboard Shortcuts</h4>
<button class="close" data-dismiss="modal" type="button"><span aria-hidden="true">×</span><span class="sr-only">Close</span></button>
</div>
<div class="modal-body">
<table class="table">
<thead>
<tr>
<th style="width: 20%;">Keys</th>
<th>Action</th>
</tr>
</thead>
<tbody>
<tr>
<td class="help shortcut"><kbd>?</kbd></td>
<td>Open this help</td>
</tr>
<tr>
<td class="next shortcut"><kbd>n</kbd></td>
<td>Next page</td>
</tr>
<tr>
<td class="prev shortcut"><kbd>p</kbd></td>
<td>Previous page</td>
</tr>
<tr>
<td class="search shortcut"><kbd>s</kbd></td>
<td>Search</td>
</tr>
</tbody>
</table>
</div>
<div class="modal-footer">
</div>
</div>
</div>
</div>
</body>
</html>
